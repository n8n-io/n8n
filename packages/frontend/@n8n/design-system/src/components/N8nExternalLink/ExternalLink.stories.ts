import type { StoryFn } from '@storybook/vue3';

import N8nExternalLink from './ExternalLink.vue';

export default {
	title: 'Atoms/ExternalLink',
	component: N8nExternalLink,
	argTypes: {
		size: {
			control: 'select',
			options: ['small', 'medium', 'large'],
		},
		newWindow: {
			control: 'boolean',
		},
	},
};

const Template: StoryFn = (args, { argTypes }) => ({
	setup: () => ({ args }),
	props: Object.keys(argTypes),
	components: {
		N8nExternalLink,
	},
	template: '<N8nExternalLink v-bind="args">{{ args.default }}</N8nExternalLink>',
});

export const IconOnly = Template.bind({});
IconOnly.args = {
	href: 'https://n8n.io',
	size: 'medium',
	newWindow: true,
};

export const WithText = Template.bind({});
WithText.args = {
	href: 'https://n8n.io',
	size: 'medium',
	newWindow: true,
	default: 'Visit n8n',
};

export const Small = Template.bind({});
Small.args = {
	href: 'https://n8n.io',
	size: 'small',
	newWindow: true,
	default: 'Visit n8n',
};

export const Large = Template.bind({});
Large.args = {
	href: 'https://n8n.io',
	size: 'large',
	newWindow: true,
	default: 'Visit n8n',
};

export const SameWindow = Template.bind({});
SameWindow.args = {
	href: 'https://n8n.io',
	size: 'medium',
	newWindow: false,
	default: 'Visit n8n',
};

export const WithClickHandler = Template.bind({});
WithClickHandler.args = {
	size: 'medium',
	onClick: () => alert('Clicked!'),
	default: 'Click me',
};
