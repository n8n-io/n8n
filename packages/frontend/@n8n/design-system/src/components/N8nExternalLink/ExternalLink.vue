<script lang="ts" setup>
import type { IconSize } from '@n8n/design-system/types';

import N8nIcon from '../N8nIcon';

interface ExternalLinkProps {
	href?: string;
	size?: IconSize;
	newWindow?: boolean;
}

defineOptions({ name: 'N8nExternalLink' });
withDefaults(defineProps<ExternalLinkProps>(), {
	href: undefined,
	size: undefined,
	newWindow: true,
});
</script>

<template>
	<component
		:is="href ? 'a' : 'button'"
		:href="href"
		:target="href && newWindow ? '_blank' : undefined"
		:rel="href && newWindow ? 'noopener noreferrer' : undefined"
		:class="$style.link"
		v-bind="$attrs"
	>
		<slot></slot>
		<N8nIcon icon="external-link" :size="size" />
	</component>
</template>

<style lang="scss" module>
.link {
	color: var(--color-text-base);
	text-decoration: none;
	display: inline-flex;
	align-items: center;
	gap: var(--spacing-4xs);
	background: none;
	border: none;
	padding: var(--spacing-4xs) var(--spacing-2xs);
	cursor: pointer;
	border-radius: var(--border-radius-base);
	font-weight: var(--font-weight-regular);

	svg {
		color: var(--color-text-light);
	}

	&:hover {
		color: var(--color-primary);
		background: var(--color-foreground-light);

		svg {
			color: var(--color-primary);
		}
	}

	&:active {
		color: var(--color-primary-shade-1);
	}
}
</style>
