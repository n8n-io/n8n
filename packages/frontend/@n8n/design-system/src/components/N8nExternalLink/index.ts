import N8nExternalLink from './ExternalLink.vue';

export default N8nExternalLink;
