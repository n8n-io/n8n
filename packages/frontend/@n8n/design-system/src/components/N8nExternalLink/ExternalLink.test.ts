import { render } from '@testing-library/vue';

import N8nExternalLink from './ExternalLink.vue';

const stubs = ['n8n-icon'];

describe('components', () => {
	describe('N8nExternalLink', () => {
		it('should render correctly with href', () => {
			const wrapper = render(N8nExternalLink, {
				props: {
					href: 'https://n8n.io',
				},
				global: {
					stubs,
				},
			});
			expect(wrapper.html()).toMatchSnapshot();
		});

		it('should render correctly without href', () => {
			const wrapper = render(N8nExternalLink, {
				props: {},
				global: {
					stubs,
				},
			});
			expect(wrapper.html()).toMatchSnapshot();
		});

		it('should render correctly with slot content', () => {
			const wrapper = render(N8nExternalLink, {
				props: { href: 'https://n8n.io' },
				slots: { default: 'Visit n8n' },
				global: {
					stubs,
				},
			});
			expect(wrapper.html()).toMatchSnapshot();
		});

		describe('props', () => {
			describe('href', () => {
				it('should set href attribute', () => {
					const href = 'https://example.com';
					const wrapper = render(N8nExternalLink, {
						props: { href },
						global: { stubs },
					});
					const link = wrapper.getByRole('link');
					expect(link.getAttribute('href')).toBe(href);
				});
			});

			describe('newWindow', () => {
				it('should open in new window by default', () => {
					const wrapper = render(N8nExternalLink, {
						props: { href: 'https://n8n.io' },
						global: { stubs },
					});
					const link = wrapper.getByRole('link');
					expect(link.getAttribute('target')).toBe('_blank');
					expect(link.getAttribute('rel')).toBe('noopener noreferrer');
				});

				it('should not open in new window when newWindow is false', () => {
					const wrapper = render(N8nExternalLink, {
						props: {
							href: 'https://n8n.io',
							newWindow: false,
						},
						global: { stubs },
					});
					const link = wrapper.getByRole('link');
					expect(link.getAttribute('target')).toBeNull();
					expect(link.getAttribute('rel')).toBeNull();
				});
			});

			describe('size', () => {
				it('should pass size prop to icon', () => {
					const wrapper = render(N8nExternalLink, {
						props: {
							href: 'https://n8n.io',
							size: 'large',
						},
						global: { stubs },
					});
					const iconStub = wrapper.html();
					expect(iconStub).toContain('size="large"');
				});
			});
		});

		describe('element type', () => {
			it('should render as anchor when href is provided', () => {
				const wrapper = render(N8nExternalLink, {
					props: { href: 'https://n8n.io' },
					global: { stubs },
				});
				const element = wrapper.getByRole('link');
				expect(element.tagName).toBe('A');
			});

			it('should render as button when href is not provided', () => {
				const wrapper = render(N8nExternalLink, {
					props: {},
					global: { stubs },
				});
				const element = wrapper.getByRole('button');
				expect(element.tagName).toBe('BUTTON');
			});
		});

		describe('slot content', () => {
			it('should display slot content before icon', () => {
				const wrapper = render(N8nExternalLink, {
					props: { href: 'https://n8n.io' },
					slots: { default: 'Visit n8n' },
					global: { stubs },
				});
				const link = wrapper.getByRole('link');
				expect(link).toHaveTextContent('Visit n8n');
			});
		});

		describe('styling', () => {
			it('should have base text color initially', () => {
				const wrapper = render(N8nExternalLink, {
					props: { href: 'https://n8n.io' },
					global: { stubs },
				});
				const link = wrapper.getByRole('link');
				expect(link).toHaveClass('link');
			});
		});
	});
});
