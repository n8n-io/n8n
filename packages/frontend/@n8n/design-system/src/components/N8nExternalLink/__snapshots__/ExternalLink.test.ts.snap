// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components > N8nExternalLink > should render correctly with href 1`] = `
"<a href="https://n8n.io" target="_blank" rel="noopener noreferrer" class="link">
  <n8n-icon-stub icon="external-link" spin="false"></n8n-icon-stub>
</a>"
`;

exports[`components > N8nExternalLink > should render correctly with slot content 1`] = `"<a href="https://n8n.io" target="_blank" rel="noopener noreferrer" class="link">Visit n8n<n8n-icon-stub icon="external-link" spin="false"></n8n-icon-stub></a>"`;

exports[`components > N8nExternalLink > should render correctly without href 1`] = `
"<button class="link">
  <n8n-icon-stub icon="external-link" spin="false"></n8n-icon-stub>
</button>"
`;
